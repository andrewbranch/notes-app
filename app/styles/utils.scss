@import './variables.scss';

// Spacing

.padding-list-xxs { padding: $space-list-xxs; }
.padding-list-xs { padding: $space-list-xs; }
.padding-list-s { padding: $space-list-s; }
.padding-list-m { padding: $space-list-m; }
.padding-list-l { padding: $space-list-l; }
.padding-list-xl { padding: $space-list-xl; }

// Borders & corners

.rounded-sm { border-radius: 2px; }

// Materials

.surface { background-color: $surface; }
.surface-selection { background-color: $surface-selection; }

// Resets

.unanchor {
  &, &:hover, &:visited, &:active {
    color: unset;
    text-decoration: unset;
    cursor: default;
  }
}