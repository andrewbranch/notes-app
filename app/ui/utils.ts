import { Styleable } from './types';

export function cn<T extends Styleable>(props: T, className: string): T {
  return Object.assign({}, props, { className: [className, props.className].join(' ') });
}