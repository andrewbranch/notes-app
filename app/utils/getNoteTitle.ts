import { ContentState, RawDraftContentState, ContentBlock, RawDraftContentBlock, EditorState } from 'draft-js';
import { collapseInlineStylesInBlock } from '../components/editor/coreStylingPlugin/steps/collapseInlineStyles';
import { performDependentEdits } from './draftUtils';

const defaultNoteTitle = 'Untitled note';

const getBlockText = (block: ContentBlock | RawDraftContentBlock) => (block instanceof ContentBlock ? block.getText() : block.text).trim();
const blockHasContents = (block: ContentBlock | RawDraftContentBlock) => !!getBlockText(block);

export const getNoteTitle = (content: ContentState | RawDraftContentState): string => {
  const blocks: (ContentBlock | RawDraftContentBlock)[] = content instanceof ContentState ? content.getBlocksAsArray() : content.blocks;
  const firstNonEmptyBlock = blocks.find(blockHasContents);
  if (firstNonEmptyBlock instanceof ContentBlock) {
    // This is lame, we need to create a dummy editor state or change `performDependentEdits`
    const dummyEditorState = EditorState.createWithContent(ContentState.createFromBlockArray([firstNonEmptyBlock]));
    return performDependentEdits(
      dummyEditorState,
      collapseInlineStylesInBlock(firstNonEmptyBlock)
    ).getCurrentContent().getFirstBlock().getText();
  }

  return firstNonEmptyBlock ? getBlockText(firstNonEmptyBlock) : defaultNoteTitle;
};
