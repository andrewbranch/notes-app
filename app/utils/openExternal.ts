import { shell } from 'electron';
import { isMacOS } from './platform';

export const openExternal = (event: React.MouseEvent<HTMLAnchorElement>) => {
  event.preventDefault();
  const anchor = event.currentTarget;
  if (anchor && 'href' in anchor) {
    shell.openExternal(anchor.href);
  }
};

export const openExternalOnCmdModifier = (event: React.MouseEvent<HTMLAnchorElement>) => {
  event.preventDefault();
  if (isMacOS && event.metaKey || !isMacOS && event.ctrlKey) {
    openExternal(event);
  }
};
