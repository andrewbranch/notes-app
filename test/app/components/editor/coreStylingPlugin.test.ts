import '../../../utils/enzymeConfig';
import { createEditorProvider, Key } from './editorProvider';
import { boldWithSelectionAtEnd, boldWithSelectionAtStart } from './fixtures';

describe('coreStylingPlugin', () => {
  describe('steps', () => {
    describe('removeInlineStyles', () => {
      describe('deleting decorator characters', () => {
        test('backspacing the first of a two-character leading decorator sequence should remove the style', async () => {
          const provider = await createEditorProvider({ initialEditorState: boldWithSelectionAtStart });
          await provider.setSelection({ focusOffset: 1, anchorOffset: 1 });
          await provider.pressKey(Key.BACKSPACE);
          expect(provider.getEditorState()).toMatchSnapshot();
        });

        test('backspacing the second of a two-character leading decorator sequence should remove the style', async () => {
          const provider = await createEditorProvider({ initialEditorState: boldWithSelectionAtStart });
          await provider.setSelection({ focusOffset: 2, anchorOffset: 2 });
          await provider.pressKey(Key.BACKSPACE);
          expect(provider.getEditorState()).toMatchSnapshot();
        });

        test('backspacing the first of a two-character trailing decorator sequence should remove the style', async () => {
          const provider = await createEditorProvider({ initialEditorState: boldWithSelectionAtEnd });
          await provider.pressKey(Key.LEFT);
          await provider.pressKey(Key.BACKSPACE);
          expect(provider.getEditorState()).toMatchSnapshot();
        });

        test('backspacing the second of a two-character trailing decorator sequence should remove the style', async () => {
          const provider = await createEditorProvider({ initialEditorState: boldWithSelectionAtEnd });
          await provider.pressKey(Key.BACKSPACE);
          expect(provider.getEditorState()).toMatchSnapshot();
        });
      });
    });
  });
});
