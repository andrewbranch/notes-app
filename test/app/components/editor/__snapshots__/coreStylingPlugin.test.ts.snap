// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`coreStylingPlugin steps removeInlineStyles deleting decorator characters backspacing the first of a two-character leading decorator sequence should remove the style 1`] = `
EditorState {
  "_immutable": Immutable.Record {
    "allowUndo": true,
    "currentContent": Immutable.Record {
      "entityMap": Object {},
      "blockMap": Immutable.OrderedMap {
        "aacj3": Immutable.Record {
          "key": "aacj3",
          "type": "unstyled",
          "text": "*Bold**",
          "characterList": Immutable.List [
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
          ],
          "depth": 0,
          "data": Immutable.Map {},
        },
      },
      "selectionBefore": Immutable.Record {
        "anchorKey": "aacj3",
        "anchorOffset": 1,
        "focusKey": "aacj3",
        "focusOffset": 1,
        "isBackward": false,
        "hasFocus": true,
      },
      "selectionAfter": Immutable.Record {
        "anchorKey": "aacj3",
        "anchorOffset": 0,
        "focusKey": "aacj3",
        "focusOffset": 0,
        "isBackward": false,
        "hasFocus": true,
      },
    },
    "decorator": MultiDecorator {
      "decorators": Immutable.List [
        CompositeDraftDecorator {
          "_decorators": Array [],
        },
      ],
    },
    "directionMap": Immutable.OrderedMap {
      "aacj3": "LTR",
    },
    "forceSelection": true,
    "inCompositionMode": false,
    "inlineStyleOverride": null,
    "lastChangeType": "change-inline-style",
    "nativelyRenderedContent": null,
    "redoStack": Immutable.Stack [],
    "selection": Immutable.Record {
      "anchorKey": "aacj3",
      "anchorOffset": 0,
      "focusKey": "aacj3",
      "focusOffset": 0,
      "isBackward": false,
      "hasFocus": true,
    },
    "treeMap": Immutable.OrderedMap {
      "aacj3": Immutable.List [
        Immutable.Record {
          "start": 0,
          "end": 7,
          "decoratorKey": null,
          "leaves": Immutable.List [
            Immutable.Record {
              "start": 0,
              "end": 7,
            },
          ],
        },
      ],
    },
    "undoStack": Immutable.Stack [
      Immutable.Record {
        "entityMap": Object {},
        "blockMap": Immutable.OrderedMap {
          "aacj3": Immutable.Record {
            "key": "aacj3",
            "type": "unstyled",
            "text": "**Bold**",
            "characterList": Immutable.List [
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                  "core.styling.decorator",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                  "core.styling.decorator",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                  "core.styling.decorator",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                  "core.styling.decorator",
                ],
                "entity": null,
              },
            ],
            "depth": 0,
            "data": Immutable.Map {},
          },
        },
        "selectionBefore": Immutable.Record {
          "anchorKey": "aacj3",
          "anchorOffset": 0,
          "focusKey": "aacj3",
          "focusOffset": 0,
          "isBackward": false,
          "hasFocus": false,
        },
        "selectionAfter": Immutable.Record {
          "anchorKey": "aacj3",
          "anchorOffset": 0,
          "focusKey": "aacj3",
          "focusOffset": 0,
          "isBackward": false,
          "hasFocus": false,
        },
      },
    ],
  },
}
`;

exports[`coreStylingPlugin steps removeInlineStyles deleting decorator characters backspacing the first of a two-character trailing decorator sequence should remove the style 1`] = `
EditorState {
  "_immutable": Immutable.Record {
    "allowUndo": true,
    "currentContent": Immutable.Record {
      "entityMap": Object {},
      "blockMap": Immutable.OrderedMap {
        "aacj3": Immutable.Record {
          "key": "aacj3",
          "type": "unstyled",
          "text": "**Bold*",
          "characterList": Immutable.List [
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
          ],
          "depth": 0,
          "data": Immutable.Map {},
        },
      },
      "selectionBefore": Immutable.Record {
        "anchorKey": "aacj3",
        "anchorOffset": 8,
        "focusKey": "aacj3",
        "focusOffset": 8,
        "isBackward": false,
        "hasFocus": true,
      },
      "selectionAfter": Immutable.Record {
        "anchorKey": "aacj3",
        "anchorOffset": 7,
        "focusKey": "aacj3",
        "focusOffset": 7,
        "isBackward": false,
        "hasFocus": true,
      },
    },
    "decorator": MultiDecorator {
      "decorators": Immutable.List [
        CompositeDraftDecorator {
          "_decorators": Array [],
        },
      ],
    },
    "directionMap": Immutable.OrderedMap {
      "aacj3": "LTR",
    },
    "forceSelection": true,
    "inCompositionMode": false,
    "inlineStyleOverride": null,
    "lastChangeType": "change-inline-style",
    "nativelyRenderedContent": null,
    "redoStack": Immutable.Stack [],
    "selection": Immutable.Record {
      "anchorKey": "aacj3",
      "anchorOffset": 7,
      "focusKey": "aacj3",
      "focusOffset": 7,
      "isBackward": false,
      "hasFocus": true,
    },
    "treeMap": Immutable.OrderedMap {
      "aacj3": Immutable.List [
        Immutable.Record {
          "start": 0,
          "end": 7,
          "decoratorKey": null,
          "leaves": Immutable.List [
            Immutable.Record {
              "start": 0,
              "end": 7,
            },
          ],
        },
      ],
    },
    "undoStack": Immutable.Stack [
      Immutable.Record {
        "entityMap": Object {},
        "blockMap": Immutable.OrderedMap {
          "aacj3": Immutable.Record {
            "key": "aacj3",
            "type": "unstyled",
            "text": "**Bold**",
            "characterList": Immutable.List [
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                  "core.styling.decorator",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                  "core.styling.decorator",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                  "core.styling.decorator",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                  "core.styling.decorator",
                ],
                "entity": null,
              },
            ],
            "depth": 0,
            "data": Immutable.Map {},
          },
        },
        "selectionBefore": Immutable.Record {
          "anchorKey": "aacj3",
          "anchorOffset": 0,
          "focusKey": "aacj3",
          "focusOffset": 0,
          "isBackward": false,
          "hasFocus": false,
        },
        "selectionAfter": Immutable.Record {
          "anchorKey": "aacj3",
          "anchorOffset": 0,
          "focusKey": "aacj3",
          "focusOffset": 0,
          "isBackward": false,
          "hasFocus": false,
        },
      },
    ],
  },
}
`;

exports[`coreStylingPlugin steps removeInlineStyles deleting decorator characters backspacing the second of a two-character leading decorator sequence should remove the style 1`] = `
EditorState {
  "_immutable": Immutable.Record {
    "allowUndo": true,
    "currentContent": Immutable.Record {
      "entityMap": Object {},
      "blockMap": Immutable.OrderedMap {
        "aacj3": Immutable.Record {
          "key": "aacj3",
          "type": "unstyled",
          "text": "*Bold**",
          "characterList": Immutable.List [
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
          ],
          "depth": 0,
          "data": Immutable.Map {},
        },
      },
      "selectionBefore": Immutable.Record {
        "anchorKey": "aacj3",
        "anchorOffset": 2,
        "focusKey": "aacj3",
        "focusOffset": 2,
        "isBackward": false,
        "hasFocus": true,
      },
      "selectionAfter": Immutable.Record {
        "anchorKey": "aacj3",
        "anchorOffset": 1,
        "focusKey": "aacj3",
        "focusOffset": 1,
        "isBackward": false,
        "hasFocus": true,
      },
    },
    "decorator": MultiDecorator {
      "decorators": Immutable.List [
        CompositeDraftDecorator {
          "_decorators": Array [],
        },
      ],
    },
    "directionMap": Immutable.OrderedMap {
      "aacj3": "LTR",
    },
    "forceSelection": true,
    "inCompositionMode": false,
    "inlineStyleOverride": null,
    "lastChangeType": "change-inline-style",
    "nativelyRenderedContent": null,
    "redoStack": Immutable.Stack [],
    "selection": Immutable.Record {
      "anchorKey": "aacj3",
      "anchorOffset": 1,
      "focusKey": "aacj3",
      "focusOffset": 1,
      "isBackward": false,
      "hasFocus": true,
    },
    "treeMap": Immutable.OrderedMap {
      "aacj3": Immutable.List [
        Immutable.Record {
          "start": 0,
          "end": 7,
          "decoratorKey": null,
          "leaves": Immutable.List [
            Immutable.Record {
              "start": 0,
              "end": 7,
            },
          ],
        },
      ],
    },
    "undoStack": Immutable.Stack [
      Immutable.Record {
        "entityMap": Object {},
        "blockMap": Immutable.OrderedMap {
          "aacj3": Immutable.Record {
            "key": "aacj3",
            "type": "unstyled",
            "text": "**Bold**",
            "characterList": Immutable.List [
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                  "core.styling.decorator",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                  "core.styling.decorator",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                  "core.styling.decorator",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                  "core.styling.decorator",
                ],
                "entity": null,
              },
            ],
            "depth": 0,
            "data": Immutable.Map {},
          },
        },
        "selectionBefore": Immutable.Record {
          "anchorKey": "aacj3",
          "anchorOffset": 0,
          "focusKey": "aacj3",
          "focusOffset": 0,
          "isBackward": false,
          "hasFocus": false,
        },
        "selectionAfter": Immutable.Record {
          "anchorKey": "aacj3",
          "anchorOffset": 0,
          "focusKey": "aacj3",
          "focusOffset": 0,
          "isBackward": false,
          "hasFocus": false,
        },
      },
    ],
  },
}
`;

exports[`coreStylingPlugin steps removeInlineStyles deleting decorator characters backspacing the second of a two-character trailing decorator sequence should remove the style 1`] = `
EditorState {
  "_immutable": Immutable.Record {
    "allowUndo": true,
    "currentContent": Immutable.Record {
      "entityMap": Object {},
      "blockMap": Immutable.OrderedMap {
        "aacj3": Immutable.Record {
          "key": "aacj3",
          "type": "unstyled",
          "text": "**Bold*",
          "characterList": Immutable.List [
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
            Immutable.Record {
              "style": Immutable.OrderedSet [],
              "entity": null,
            },
          ],
          "depth": 0,
          "data": Immutable.Map {},
        },
      },
      "selectionBefore": Immutable.Record {
        "anchorKey": "aacj3",
        "anchorOffset": 8,
        "focusKey": "aacj3",
        "focusOffset": 8,
        "isBackward": false,
        "hasFocus": true,
      },
      "selectionAfter": Immutable.Record {
        "anchorKey": "aacj3",
        "anchorOffset": 7,
        "focusKey": "aacj3",
        "focusOffset": 7,
        "isBackward": false,
        "hasFocus": true,
      },
    },
    "decorator": MultiDecorator {
      "decorators": Immutable.List [
        CompositeDraftDecorator {
          "_decorators": Array [],
        },
      ],
    },
    "directionMap": Immutable.OrderedMap {
      "aacj3": "LTR",
    },
    "forceSelection": true,
    "inCompositionMode": false,
    "inlineStyleOverride": null,
    "lastChangeType": "change-inline-style",
    "nativelyRenderedContent": null,
    "redoStack": Immutable.Stack [],
    "selection": Immutable.Record {
      "anchorKey": "aacj3",
      "anchorOffset": 7,
      "focusKey": "aacj3",
      "focusOffset": 7,
      "isBackward": false,
      "hasFocus": true,
    },
    "treeMap": Immutable.OrderedMap {
      "aacj3": Immutable.List [
        Immutable.Record {
          "start": 0,
          "end": 7,
          "decoratorKey": null,
          "leaves": Immutable.List [
            Immutable.Record {
              "start": 0,
              "end": 7,
            },
          ],
        },
      ],
    },
    "undoStack": Immutable.Stack [
      Immutable.Record {
        "entityMap": Object {},
        "blockMap": Immutable.OrderedMap {
          "aacj3": Immutable.Record {
            "key": "aacj3",
            "type": "unstyled",
            "text": "**Bold**",
            "characterList": Immutable.List [
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                  "core.styling.decorator",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                  "core.styling.decorator",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                  "core.styling.decorator",
                ],
                "entity": null,
              },
              Immutable.Record {
                "style": Immutable.OrderedSet [
                  "BOLD",
                  "core.styling.decorator",
                ],
                "entity": null,
              },
            ],
            "depth": 0,
            "data": Immutable.Map {},
          },
        },
        "selectionBefore": Immutable.Record {
          "anchorKey": "aacj3",
          "anchorOffset": 0,
          "focusKey": "aacj3",
          "focusOffset": 0,
          "isBackward": false,
          "hasFocus": false,
        },
        "selectionAfter": Immutable.Record {
          "anchorKey": "aacj3",
          "anchorOffset": 0,
          "focusKey": "aacj3",
          "focusOffset": 0,
          "isBackward": false,
          "hasFocus": false,
        },
      },
    ],
  },
}
`;
